Move specialist document titles into the document model, simplify ApplicationController

benilovj opened the pull request:
This means that more of the format-specific logic lives in the document
model, not centrally.

This PR also includes:
- a correction to 2 document format titles (EAT Decisions, ET decisions) from plural to singular, to fit the pattern of the other formats.
- a simplification of `ApplicationController` to reduce unnecessary indirection

benilovj committed:
Move specialist document titles into the document model

This means that more of the format-specific logic lives in the document
model, not centrally.

I've also corrected 2 document titles (EAT Decisions, ET decisions)
from plural to singular, to fit the pattern of the other formats.

h-lame commented on app/controllers/application_controller.rb:
```diff
@@ -42,48 +42,44 @@ def formats_user_can_access
+          document_class.organisations,
         )
       }
     }.reduce({}, :merge)
```
While we're tidying this up how about also:

1) memoizing it; this method isn't free to run and is probably used a bunch
2) getting rid of the `map` + `reduce` dance by using `.map.with_object`:

``` ruby
document_classes.map.with_object({}) do |document_class, everything|
  title = document_class.title
  thing = # build the FormatStruct instance
  everything[thing.document_type] = thing
end
```

When I read the `map` block and see it's followed by the `reduce` I have to spend a moment working out what it's doing and why, whereas it'd be upfront in the `map.with_object` case.  YMMV obvs.

I'm more interested in memoizing it than this TBH.

benilovj replied:
I've done away with the entire `FormatStruct` business in 3ac24f8. Do you think memoizing is still necessary?

benilovj committed:
Simplify ApplicationsController

This commit gets rid of FormatStruct completely, which was not doing
much more than present a few simple text transformations from Document.
Those methods have been moved to Document directly, and so Document
classes can be used directly without any wrapper. This simplifies
the controller and view code that consumes the document types.

h-lame commented on app/controllers/application_controller.rb:
```diff
@@ -26,66 +26,27 @@ def document_type
   def current_format
-    document_types.fetch(params.fetch(:document_type, nil), nil)
+    document_types.detect { |format| format.document_type == document_type }
   end
```
Old `current_format` impl. was a hash lookup so not much to be gained by memoisation, but now it's iterating through an array to find something, so there will be tiny benefits to be gained for all requests for `VehicleRecallsAndFaultsAlert` documents ;)

benilovj replied:
done

h-lame commented on app/controllers/application_controller.rb:
```diff
@@ -26,66 +26,27 @@ def document_type
-    document_types.select { |_, v| policy(v.klass).index? }
+    document_types.select { |format| policy(format).index? }
   end
 
```
Might not hurt to memoize this while we're at it too (not sure how much it's used though, if it's just in the drop down for the header maybe it's of less use).

benilovj replied:
I don't think it's worth doing here, as it's only used in the dropdown.

h-lame replied:
Fair enough.
